### DISCUSSION TITLE
How to perform morphological operations on 3D image via porespy fftmorphology?

### DISCUSSION THREAD

**Discussion by Camelot18** [PMEAL/porespy#620] [Q&A] [answered]
Dear Professor Gostick, 

I am a Ph.D. student from Peking University, interested in pore network modeling, and I'm trying to construct a 3D pore network with different hydrate saturation and morphologies, which will be imported into Comsol for FEM simulation. But there are a few errors that occurred during my attempts.
(1) I followed the examples to extract the pore network and want to plot the connections and coordinates on the same graph, but it either shows "'Line3DCollection' object has no property 'ax'" or "inner() got multiple values for argument 'ax'". I can only plot them separately. The codes I used are as below.

fig, ax = plt.subplots(1,1,figsize=[8, 8])
ax.imshow(graphx, cmap=plt.cm.magma);
ax = op.topotools.plot_coordinates(ax=ax,
                              network=net,
                              size_by=net["pore.inscribed_diameter"],
                              color_by=net["pore.inscribed_diameter"],
                              markersize=200)
ax = op.topotools.plot_connections(network=net,
                                   size_by=net["throat.inscribed_diameter"],
                              color_by=net["throat.inscribed_diameter"],ax=ax)
![image](https://user-images.githubusercontent.com/94626894/150317453-1c99a891-4bd6-4544-8ac9-f2783ce1c3d7.png)
![image](https://user-images.githubusercontent.com/94626894/150317535-3769ba76-e1af-4de8-95a8-217a77574684.png)
![image](https://user-images.githubusercontent.com/94626894/150317558-7d234ac8-d068-4863-bc34-0595bb52ece4.png)

(2) I also want to calculate the tortuosity of the pore networks, but it didn't give me a correct value. Do you have any examples to do that for 3D images?

(3)Last step is the most challenging one for me, I need to distribute gas hydrate(solid phase) into the constructed pore network with different locations to generate various morphologies(cementing, grain contacting, and pore-filling) via morphological operations(erosion, dilation). But for 3D pore network, it seems that the fftmorphology in porespy doesn't work for me. I also need to control the hydrate content(effective porosity would be a controlling parameter), how can I create such pore networks with a specific hydrate saturation and morphology?
Your reply will be greatly appreciated!

Thanks in advance,
Peter

**Comment by jgostick** [comment, marked as answer]
Hi @Camelot18 
That's a lot going on, let me try to address it in order:

1) For plotting large networks, use Paraview.  We are in the middle of reorganizing our docs, and also releasing V3, and it's taking ages (sorry!).  We realized that somehow the paraview tutorial got lost.  I have uploaded it [here temporarily](https://github.com/PMEAL/OpenPNM/blob/readd_paraview/examples/notebooks/io/view_vtp_in_paraview.md).  Trying to plot all these together in matplotlib (which is what ``plot_connections`` etc uses, is a bit more than it can handle.

2) Tortuosity of the network comes from computing the effective diffusivity using the FickianDiffusion class, then normalizing by the open-space diffusivity and porosity accordingly.  This means you need to get the effective diffusivity correct.  The ``snow`` network extraction usually does a decent job but sometime you need to override the properties with ones that make more sense for your material.  I'm going to direct you you a unit test we have that does this for Berea sandstone which may help: https://github.com/PMEAL/OpenPNM/blob/readd_paraview/tests/integration/PoreSpyIO_on_Berea.py

3) If ``fftmorphology`` does not work, then just try using the standard ``scipy.ndimage.binary_dilation/erosion/etc``.  They may be more robust.  I assume you're adding these hydrates before you extract the network?  If you want to add them to the network after extraction, then the image processing tools no longer apply and you should use other logic to figure out which pores are filled with hydrates, like percolation, or some other physics-based algorithm.

**Reply by Camelot18** [reply]
Dear professor,

Thank you for your quick reply which is really helpful.
I tried paraview and it worked.
![a284a7c585b6d3dd7d3f6db5ac6931a](https://user-images.githubusercontent.com/94626894/150566158-0f41c1b7-1338-4ebd-93ee-468b123ab0d2.png)


As for the Tortuosity calculation, I was using ps.dns.tortuosity given by https://porespy.org/_modules/porespy/dns/_funcs.html, but it returned with "_MemoryError: Unable to allocate 277. MiB for an array with shape (72739423,) and data type int32_". How can I resolve this? Should I calculate effective diffusivity and follow the steps you provided?
The code I used is as below.
tau_1 = ps.dns.tortuosity(im, axis=1)
tau_3 = ps.dns.tortuosity(im, axis=3)
print(f"Horizontal tortuosity: {tau_1:.2f}%")
print(f"Vertical tortuosity: {tau_3:.2f}%")

In terms of adding hydrate into the pore space, the idea is to quantitatively put it into the specific positions(pore center, pore surface, and throat corresponding to the schematic below).
![image](https://user-images.githubusercontent.com/94626894/150566301-c31b755d-62af-49ea-a5f2-780a882ff0b2.png)


Some proposed morphological operations to generate such pore networks(see below), I am kind of struggling with realizing this method. Do you have any suggestions?
![image](https://user-images.githubusercontent.com/94626894/150566236-c639e730-87bd-4760-8b3a-46054c9d3928.png)

Looking forward to your answers.

Best,
Peter

**Reply by jgostick** [reply]
The dns approach only works if the image is ~100-150 cubed, unless you have a super duper computer. There is a matlab package called taufactor that works the same way.  They may be able to do bigger images, but also will have problems.  Apparently the GUI estimates how long it will take before you run it.  When the image is large you can consider doing a random walk, which developed pytrax for.

**Reply by jgostick** [reply]
RE: the hydrates, it looks they are basically growing and blocking the pores and throats.  You can do this in the PNM by choosing sites for nucleation the 'growing' the hydrates by selecting the neighbors of these sites and so on until your hydrate crystal is the size you want.  Then you would reduce the diffusive conductance of those pores and throats using some analytical expression which you'll derive based on the type of growth (float, cemented, etc).

**Reply by Camelot18** [reply]
Thanks for your kind responses.
Yes, the image is 300 cubed. I will try to use pytrax for the tortuosity calculation.
For hydrate growth, are you referring to the diffusion-limited aggregation (DLA) model? Should I assign the identical geometry for hydrate? Still pondering how to perform this.

**Reply by jgostick** [reply]
For the hydrate model, you're going to need to basically figure it out.  If there has been another paper on hydrate growth in PNMs, then you can follow that.  If not then you're blazing new territory!

**Reply by Camelot18** [reply]
I think I am kind of figured it out, will try if it works.
In OpenPNM package, can I do the subtract operation? For pore-filling hydrate, I need to delete/remove the hydrate(treated as a sphere) part in the pore space to generate a new pore network.
Besides, how should I export the newly constructed pore network as a comsol file(geometry) for the later simulation?

Thanks,
Peter

**Reply by jgostick** [reply]
You can 'delete pores' using ``openpnm.topotools.trim`` but that is not the normal way to do it.  Normally you'd set transport conductances of those pores and their adjacent throats to negligible value.

**Reply by Camelot18** [reply]
Dear professor,

Thank you for your kind reply.
Since the radius of the hydrate(although smaller than the pore radius) is to be determined by the saturation and the position could be random in the pore, which means "deleting the pore" is not correct. Rather, I need to "put" the hydrate into the pore first and then subtract this part by pore in order to obtain a new network(needs to be visualized and analyzed) corresponding to the hydrate-bearing samples with a specific saturation. Is it possible to export these 3D pore networks into Comsol format or should I perform the FEM simulation within python?

Best,
Peter

**Reply by jgostick** [reply]
There is nothing quite as simple as a subtraction, but you can recompute the pore volume and throat diameters based on the size of the hydrates, then you can recompute the transport conductance, which will be reduced by the presence of the solid.

**Reply by Camelot18** [reply]
Yes, like in Comsol I can just perform the "difference" operation for the two objects. Hydrate can be present in a random position of pores, which will produce a different pore network. I can recalculate the pore volume, but the new network connectivity can only be visualized by this sort of operation. 
Could you also answer my second question that if I can export 3D pore network in Comsol format? Openpnm only supports 2D network while porespy handles with voxel image.
